### PR TITLE
[RFC] Configure Cuda options if targets starts with nvptx

### PR DESCRIPTION
**Pull request by vchuravy**
```julia
CI = Cxx.new_clang_instance(false, false, target = "nvptx64-nvidia-c    uda", CPU = "sm_35")
addHeaderDir(CI, "/opt/cuda/include", kind = C_System)

# Needs at least LLVM_VER=3.8 
cxxinclude(CI, "__clang_cuda_runtime_wrapper.h")
```

I think I still miss some options, but this gets me 95% of the way.

### CODE REVIEW

File: src/bootstrap.cpp [hunk 1]
```diff
@@ -1303,6 +1305,16 @@ static void set_default_clang_options(C, bool CCompiler, const char *Triple, con
     }
     Cxx->CI->getLangOpts().ImplicitInt = 0;
     Cxx->CI->getLangOpts().PICLevel = 2;
+    if (isnvptx) {
+        Cxx->CI->getLangOpts().CUDA = 1;
+        Cxx->CI->getLangOpts().CUDAIsDevice = 1;
+        Cxx->CI->getLangOpts().CPlusPlus14 = 0;
```
Comment (Keno) [inline src/bootstrap.cpp]: Why is this?

File: src/bootstrap.cpp [hunk 2]
```diff
@@ -1283,6 +1283,8 @@ static void set_default_clang_options(C, bool CCompiler, const char *Triple, con
 {
     T_pvalue_llvmt = _T_pvalue_llvmt;
 
+    bool isnvptx = Triple != NULL && std::string(Triple).compare(0,5,"nvptx") == 0;
```
Comment (Keno) [inline src/bootstrap.cpp]: I'd prefer to do this by letting `llvm::Triple` parse it and checking the enum value.

File: src/bootstrap.cpp (hunk 1 above)
Comment (vchuravy) [inline src/bootstrap.cpp]: CUDA only recently started to support Cxx11, so I thought it prudent to deactivate cxx14

Comment (Keno) [inline src/bootstrap.cpp]: The problem is that Cxx.jl relies on some C++14 (language, not library) features for proper operation, so I wouldn't disable this unless there are actual problems that you encounter.

Comment (vchuravy) [inline src/bootstrap.cpp]: Alright!

**Comment by vchuravy** [comment]
thanks!

File: src/bootstrap.cpp [hunk 3]
```diff
@@ -1303,6 +1308,15 @@ static void set_default_clang_options(C, bool CCompiler, const char *Triple, con
     }
     Cxx->CI->getLangOpts().ImplicitInt = 0;
     Cxx->CI->getLangOpts().PICLevel = 2;
+    if (isnvptx) {
+        Cxx->CI->getLangOpts().CUDA = 1;
+        Cxx->CI->getLangOpts().CUDAIsDevice = 1;
+        Cxx->CI->getLangOpts().DeclSpecKeyword = 1;
+#ifdef LLVM38
```
Comment (Keno) [inline src/bootstrap.cpp]: These don't seem to exist in LLVM 3.9 anymore.

Comment (Keno) [inline src/bootstrap.cpp]: https://github.com/llvm-mirror/clang/commit/b713fab2f5edb1492e056eeea54934477e4158a2 will disable on 3.9

Comment (vchuravy) [inline src/bootstrap.cpp]: while you at ti can you enable `CUDADisableTargetCallChecks` for 3.8?

**Pull request merged**